metrics/cli/lib/cli/updater/algorithm/update_algorithm.dart

```diff
@@ -0,0 +1,191 @@
+  /// Throws an [ArgumentError] if the given [fileHelper] is `null`.
+  UpdateAlgorithm({
+    @required Services services,
+    @required FileHelper fileHelper,
+  })  : _gitService = services?.gitService,
+        _npmService = services?.npmService,
+        _flutterService = services?.flutterService,
+        _sentryService = services?.sentryService,
+        _firebaseService = services?.firebaseService,
+        _fileHelper = fileHelper {
+    ArgumentError.checkNotNull(services, 'services');
+    ArgumentError.checkNotNull(_fileHelper, 'fileHelper');
+  }
+
+  /// Starts the updating of the deployed Metrics components using the
+  /// given [config] and [paths].
+  ///
+  /// Throws an [ArgumentError] if the given [config] is `null`.
+  /// Throws an [ArgumentError] if the given [paths] is `null`.
+  Future<void> start(UpdateConfig config, Paths paths) async {
+    ArgumentError.checkNotNull(config, 'config');
+    ArgumentError.checkNotNull(paths, 'paths');
+
+    await _gitService.checkout(DeployConstants.repoURL, paths.rootPath);
+    await _installNpmDependencies(
+      paths.firebasePath,
+      paths.firebaseFunctionsPath,
+    );
+    await _flutterService.build(paths.webAppPath);
+
+    final firebaseConfig = config.firebaseConfig;
+
+    _firebaseService.initializeAuth(firebaseConfig.authToken);
+
+    final sentryWebConfig = await _setupSentry(
+      config.sentryConfig,
+      paths.webAppPath,
+      paths.webAppBuildPath,
+    );
+    final metricsConfig = WebMetricsConfig(
+      googleSignInClientId: firebaseConfig.googleSignInClientId,
+      sentryWebConfig: sentryWebConfig,
+    );
+
+    _applyMetricsConfig(metricsConfig, paths.metricsConfigPath);
+
+    await _deployToFirebase(
+      firebaseConfig.projectId,
+      paths.firebasePath,
+      paths.webAppPath,
+    );
+
+    _firebaseService.resetAuth();
+  }
+
+  /// Installs npm dependencies within the given [firebasePath] and
+  /// the [functionsPath].
+  Future<void> _installNpmDependencies(
+    String firebasePath,
+    String functionsPath,
+  ) async {
+    await _npmService.installDependencies(firebasePath);
+    await _npmService.installDependencies(functionsPath);
+  }
+
+  /// Sets up a Sentry for the application under deployment within
+  /// the given [webPath] and the [buildWebPath].
+  Future<SentryWebConfig> _setupSentry(
+    SentryConfig config,
+    String webPath,
+    String buildWebPath,
+  ) async {
+    if (config == null) return null;
+
+    _sentryService.initializeAuth(config.authToken);
+
+    final release = config.releaseName;
+    final sentryProject = SentryProject(
+      organizationSlug: config.organizationSlug,
+      projectSlug: config.projectSlug,
+    );
+    final sentryRelease = SentryRelease(
+      name: release,
+      project: sentryProject,
+    );
+
+    await _createSentryRelease(
+      sentryRelease,
+      webPath,
+      buildWebPath,
+    );
+
+    _sentryService.resetAuth();
```

Comment:
🏗️ 
What if the service throws before the reset auth is called? The service authentication seems to be never reset then. It seems like we should handle this case for the firebase service as well.